server/src/main/scala/org/http4s/armeria/server/ArmeriaHttp4sHandler.scala

```diff
@@ -22,43 +26,48 @@ import com.linecorp.armeria.common.{
 }
 import com.linecorp.armeria.common.util.Version
 import com.linecorp.armeria.server.{HttpService, ServiceRequestContext}
-import io.chrisdavenport.vault.{Key => VaultKey, Vault}
+import org.typelevel.vault.{Key => VaultKey, Vault}
 import fs2._
 import fs2.interop.reactivestreams._
 import java.net.InetSocketAddress
+
 import org.http4s.internal.CollectionCompat.CollectionConverters._
 import ArmeriaHttp4sHandler.{RightUnit, canHasBody, defaultVault, toHttp4sMethod}
+import cats.effect.std.Dispatcher
+import com.comcast.ip4s.SocketAddress
 import org.http4s.server.{
   DefaultServiceErrorHandler,
   SecureSession,
   ServerRequestKeys,
   ServiceErrorHandler
 }
+import org.typelevel.ci.CIString
 import scodec.bits.ByteVector
 
 /** An [[HttpService]] that handles the specified [[HttpApp]] under the specified `prefix`. */
 private[armeria] class ArmeriaHttp4sHandler[F[_]](
     prefix: String,
     service: HttpApp[F],
-    serviceErrorHandler: ServiceErrorHandler[F]
-)(implicit F: ConcurrentEffect[F])
+    serviceErrorHandler: ServiceErrorHandler[F],
+    dispatcher: Dispatcher[F]
+)(implicit F: Async[F])
     extends HttpService {
 
-  val prefixLength: Int = if (prefix.endsWith("/")) prefix.length - 1 else prefix.length
+  val prefixLength: Int = Uri.Path.unsafeFromString(prefix).segments.size
   // micro-optimization: unwrap the service and call its .run directly
   private val serviceFn: Request[F] => F[Response[F]] = service.run
 
   override def serve(ctx: ServiceRequestContext, req: HttpRequest): HttpResponse = {
     val responseWriter = HttpResponse.streaming()
-    F.runAsync(
+    dispatcher.unsafeRunAndForget(
```

Comment:
It is shame that we could not execute a service in a context-aware event loop.
I tried to wrap the handler with `F.evalOn(x, ExecutionContext.fromExecutorService(ctx.eventLoop()))`,
the multipart test suite was not finished infinitely. 
https://github.com/http4s/http4s-armeria/blob/c8db912725f88b386135cdf4b40ffcb002b53a92/server/src/test/scala/org/http4s/armeria/server/ArmeriaServerBuilderSuite.scala#L128
I debugged but was hard to figure why it was stuck. That might be a deadlock between input streaming and output stream.

If it is impossible at the moment, we could fix the following tests?
https://github.com/http4s/http4s-armeria/blob/c8db912725f88b386135cdf4b40ffcb002b53a92/server/src/test/scala/org/http4s/armeria/server/ArmeriaServerBuilderSuite.scala#L77-L97

Reply:
I initially tried doing `unsafeRunSync` here but that also caused a deadlock. To be honest I don't really understand how the existing code succeeds in running this on the event loop. I was hoping you'd have a solution, but if not then I'll just disable those two tests.